src/OpenLoco/src/Vehicles/VehicleBogie.cpp

```diff
@@ -119,6 +104,164 @@ namespace OpenLoco::Vehicles
+                int32_t sin_angle = Math::Trigonometry::kYawToDirectionVector[this->spriteYaw].y;
+                int32_t x_distance = (distance * cos_angle) / 4096;
+                int32_t y_distance = (distance * sin_angle) / 4096;
+
+                int16_t x_distance_low = x_distance & 0x0000FFFF;
+                int16_t y_distance_low = y_distance & 0x0000FFFF;
+                int16_t x_distance_hi = x_distance / 65536;
+                int16_t y_distance_hi = y_distance / 65536;
+                int32_t newTileX = this->tileX + x_distance_low;
+                if (newTileX >= 0x00010000)
+                {
+                    x_distance_hi += 1; // carry bit
+                }
+                this->tileX = newTileX & 0x0000FFFF;
+
+                int32_t newTileY = this->tileY + y_distance_low;
+                if (newTileY >= 0x00010000)
+                {
+                    y_distance_hi += 1; // carry bit
+                }
+                this->tileY = newTileY & 0x0000FFFF;
+
+                this->tileBaseZ++;
+                int16_t zDistance = this->tileBaseZ / 32;
+
+                // Calculate new position - but don't update yet!! This is pushed to the stack.
+                World::Pos3 newPosition{ this->position.x + x_distance_hi, this->position.y + y_distance_hi, this->position.z - zDistance };
+
+                if (!sub_4AA959(this->position))
+                {
+                    World::Pos3 position_to_test = { newPosition.x, newPosition.y, this->position.z };
+                    if (sub_4AA959(position_to_test))
+                    {
+                        newPosition.x = this->position.x;
+                        newPosition.y = this->position.y;
+
+                        if (this->var_5A >= 0x50000)
+                        {
+                            rotateAndExplodeIfNotAlreadyExploded();
+                        }
+
+                        this->var_5A = ((this->var_5A & 0x7FFFFFFF) / 2) | 0x80000000;
+                    }
+
+                    if (sub_4AA959(newPosition))
+                    {
+                        newPosition.z = this->position.z;
+                        if (this->tileBaseZ >= 0x0A)
+                        {
+                            rotateAndExplodeIfNotAlreadyExploded();
+                        }
+
+                        this->tileBaseZ = 0;
+                    }
+                }
+
+                World::Pos2 currentPosition2D{ newPosition.x, newPosition.y };
+                auto newTileHeight = World::TileManager::getHeight(currentPosition2D);
+                if (newTileHeight.landHeight >= this->position.z
+                    || newTileHeight.landHeight >= newPosition.z)
+                {
+                    if (newTileHeight.landHeight < this->position.z
+                        && newTileHeight.landHeight >= newPosition.z)
+                    {
+                        newTileHeight.landHeight = this->position.z;
```

Comment:
`newTileHeight.landHeight` should be const this wouldn't be setting the value its just setting `newPosition.z`